client/materialView.ts

```diff
@@ -0,0 +1,63 @@
+import * as cg from 'chessgroundx/types';
+import * as util from 'chessgroundx/util';
+
+import { h, VNode } from 'snabbdom';
+
+import { patch } from './document';
+import { calculateGameImbalance, equivalentLetter, MaterialDiff } from './material';
+import { RoundController } from './roundCtrl';
+
+function generateContent(ctrl: RoundController, imbalance: MaterialDiff, color: cg.Color): VNode[] {
+    const result : VNode[] = [];
+
+    const keys = [...ctrl.variant.pieceRoles(util.opposite(color))]
+    for (const piece of ctrl.variant.pieceRoles(color)) {
+        const p = equivalentLetter(ctrl.variant, piece);
+        if (!keys.includes(p))
+            keys.push(p);
+    }
+    if (ctrl.variant.promotion === 'shogi') {
+        for (const piece of ctrl.variant.promoteablePieces) {
+            const p = equivalentLetter(ctrl.variant, piece);
+            if (!keys.includes(p))
+                keys.push(p);
+        }
+    }
+
+    for (const letter of keys) {
+        const p = equivalentLetter(ctrl.variant, letter);
+        const pieceDiff = (imbalance.get(p) ?? 0) * (color === 'white' ? -1 : 1);
+        if (pieceDiff > 0) {
+            const current_div : VNode[] = [];
+            for (let i = 0; i < pieceDiff; i++)
+                current_div.push(h('mpiece.' + p))
+            result.push(h('div', current_div));
+        }
+    }
+    return result;
+}
+
+function makeMaterialVNode(ctrl: RoundController, position: 'top'|'bottom', color: cg.Color, content: VNode[], disabled = false): VNode {
+    return h(`div.material.material-${position}.${color}.${ctrl.variant.piece}${disabled ? '.disabled' : ''}`, content);
+}
+
+export function updateMaterial(ctrl: RoundController, vmaterial0?: VNode | HTMLElement, vmaterial1?: VNode | HTMLElement) {
```

Comment:
Does this function still need vmaterial parameters anymore?

Reply:
Good point. Removed.